crates/router/src/core/payment_methods/cards.rs

```diff
@@ -216,6 +225,208 @@ pub async fn get_or_insert_payment_method(
+    } else {
+        let payment_method_id = generate_id(consts::ID_LENGTH, "pm");
+
+        let res = create_payment_method(
+            db,
+            &req,
+            customer_id.as_str(),
+            payment_method_id.as_str(),
+            None,
+            merchant_id.as_str(),
+            None,
+            None,
+            None,
+            key_store,
+            None,
+            Some(enums::PaymentMethodStatus::AwaitingData),
+        )
+        .await?;
+
+        Ok(services::api::ApplicationResponse::Json(
+            api::PaymentMethodResponse {
+                merchant_id: merchant_id.to_string(),
+                customer_id: Some(customer_id),
+                payment_method_id,
+                payment_method: req.payment_method,
+                payment_method_type: req.payment_method_type,
+                card: None,
+                recurring_enabled: false,
+                installment_payment_enabled: false,
+                payment_experience: None,
+                metadata: req.metadata,
+                created: None,
+                bank_transfer: None,
+                last_used_at: None,
+                client_secret: res.client_secret,
+            },
+        ))
+    }
+}
+
+#[instrument(skip_all)]
+pub async fn authenticate_pm_client_secret(
+    req_client_secret: &String,
+    db: &dyn db::StorageInterface,
+    pm_id: &str,
+) -> errors::CustomResult<diesel_models::PaymentMethod, errors::ApiErrorResponse> {
+    let payment_method = db
+        .find_payment_method(pm_id)
+        .await
+        .change_context(errors::ApiErrorResponse::PaymentMethodNotFound)
+        .attach_printable("Unable to find payment method")?;
+
+    let stored_client_secret = payment_method
+        .client_secret
+        .clone()
+        .get_required_value("client_secret")
+        .change_context(errors::ApiErrorResponse::PaymentMethodNotFound)
+        .attach_printable("client secret not found in db")?;
+
+    if req_client_secret != &stored_client_secret {
+        Err((errors::ApiErrorResponse::ClientSecretInvalid).into())
+    } else {
+        let current_timestamp = common_utils::date_time::now();
+        let session_expiry = payment_method
+            .created_at
+            .saturating_add(time::Duration::seconds(consts::DEFAULT_SESSION_EXPIRY));
+
+        fp_utils::when(current_timestamp > session_expiry, || {
+            Err::<(), errors::ApiErrorResponse>(errors::ApiErrorResponse::ClientSecretExpired)
+        })?;
+
+        Ok(payment_method)
+    }
+}
+
+#[instrument(skip_all)]
+pub async fn add_payment_method_data(
+    state: routes::AppState,
+    req: api::PaymentMethodCreate,
+    merchant_account: domain::MerchantAccount,
+    key_store: domain::MerchantKeyStore,
+    pm_id: String,
+) -> errors::RouterResponse<api::PaymentMethodResponse> {
+    let db = &*state.store;
+    let customer_id = req.customer_id.clone().get_required_value("customer_id")?;
```

Comment:
We should not be collecting `customer_id` using `client_secret`. `customer_id` should be sent in the `create_payment_method` by the merchant, use the same `customer_id` here